pkg/controller/controller.go

```diff
@@ -108,25 +106,97 @@ func (sc *Controller) performDataMigration(schedule []*schema.Schedule) error {
 			// Create migration task only if there are intersections
 			if len(intersections) > 0 {
 				key := fmt.Sprintf("%d->%d", oldShardID, newShardID)
-				if _, exists := migrations[key]; !exists {
-					migrations[key] = &MigrationTask{
-						fromShardInfo: shards[0],
-						toShardInfo:   shards[1],
-						ranges:        intersections,
-					}
+				if _, exists := ops[key]; !exists {
+					ops[key] = NewOperator(shards[0], shards[1], intersections, nil)
 				}
 			}
 		}
 	}
-
 	// Consolidate overlapping ranges for each migration
-	for _, task := range migrations {
-		task.ranges = consolidateKeyRanges(task.ranges)
+	for _, op := range ops {
+		op.keyRanges = schema.ConsolidateKeyRanges(op.keyRanges)
+		sc.operators = append(sc.operators, op)
+
+		ctx, cancel := context.WithTimeout(context.Background(), schema.MIGRATION_TIMEOUT)
+		op.cancel = cancel
+		op.done = make(chan struct{})
+
+		// Launch migration in goroutine
+		err := op.migrate(ctx)
+
+		// Wait with timeout
+		if err != nil {
+			log.Printf("Migration failed: %v", err)
+		} else {
+			log.Printf("Migration completed")
+		}
+
+		go func(op *Operator) {
+			select {
+			case <-op.done:
+				log.Printf("Migration finished between shards %d and %d", op.fromShard.ShardID, op.toShard.ShardID)
+			case <-ctx.Done():
+				log.Printf("Migration timed out, initiating cancellation")
+				if err := op.cancelMigration(); err != nil {
+					log.Printf("Failed to cancel migration: %v", err)
+				}
+			}
+		}(op)
+	}
+	return nil
+}
+
+func (sc *Controller) findOperator(fromClusterID, toClusterID uint64) *Operator {
+	for _, op := range sc.operators {
+		if op.fromShard.ShardID == fromClusterID && op.toShard.ShardID == toClusterID {
+			return op
+		}
 	}
+	return nil
+}
 
+func (sc *Controller) findOperatorByTaskID(taskId uint64) *Operator {
+	for _, op := range sc.operators {
+		if op.taskId == taskId {
+			return op
+		}
+	}
 	return nil
 }
 
+func (sc *Controller) CancelMigrationFromNode(ctx context.Context, req *pb.CancelMigrationRequest) (*pb.CancelMigrationResponse, error) {
+	log.Println("Received request to cancel migration")
+	op := sc.findOperator(req.GetFromClusterID(), req.GetToClusterID())
+	if op == nil {
+		log.Panicln("Migration not found")
+	}
+	op.cancelMigration()
+
+	return &pb.CancelMigrationResponse{
```

Comment:
Minor issue: returning "OK" without checking the result of `op.cancelMigration()`